rpc_client/src/client.rs

```diff
@@ -0,0 +1,330 @@
+use std::mem;
+use std::time;
+
+use std::path::PathBuf;
+use util::Hashable;
+use url::Url;
+use std::fs::File;
+
+use ws::{connect,
+		 Request,
+		 Handler,
+		 Sender,
+		 Handshake,
+		 Error as WsError,
+		 ErrorKind as WsErrorKind,
+		 Message,
+		 Result as WsResult};
+
+use serde::Serialize;
+use serde::Deserialize;
+use serde::ser::Serializer;
+use serde_json::{from_str,
+				 to_string,
+				 from_value,
+				 Value as JsonValue,
+				 Error as JsonError};
+
+use futures::{BoxFuture, Canceled, Complete, Future, oneshot, done};
+
+/// The actual websocket connection handler, passed into the
+/// event loop of ws-rs
+struct RpcHandler {
+	pending: Pending,
+	// Option is used here as
+	// temporary storage until
+	// connection is setup
+	// and the values are moved into
+	// the new `Rpc`
+	complete: Option<Complete<Result<Rpc, RpcError>>>,
+	auth_code: String,
+	out: Option<Sender>,
+}
+
+impl RpcHandler {
+	fn new(out: Sender,
+		   auth_code: String,
+		   complete: Complete<Result<Rpc, RpcError>>)
+		   -> Self {
+		RpcHandler {
+			out: Some(out),
+			auth_code: auth_code,
+			pending: Pending::new(),
+			complete: Some(complete),
+		}
+	}
+}
+
+impl Handler for RpcHandler {
+    fn build_request(&mut self, url: &Url) -> WsResult<Request> {
+		match Request::from_url(url) {
+			Ok(mut r) => {
+				let timestamp = time::UNIX_EPOCH.elapsed().unwrap().as_secs();
+				let hashed = format!("{}:{}", self.auth_code, timestamp).sha3();
+				let proto = format!("{:?}_{}", hashed, timestamp);
+				r.add_protocol(&proto);
+				Ok(r)
+			},
+			Err(e) => Err(WsError::new(WsErrorKind::Internal, format!("{}", e))),
+		}
+    }
+    fn on_error(&mut self, err: WsError) {
+		match mem::replace(&mut self.complete, None) {
+			Some(c) => c.complete(Err(RpcError::WsError(err))),
+			None => println!("warning: unexpected error"),
+		}
+    }
+    fn on_open(&mut self, _: Handshake) -> WsResult<()> {
+		match mem::replace(&mut self.complete, None) {
+			Some(c) => c.complete(Ok(Rpc {
+				out: mem::replace(&mut self.out, None).unwrap(),
+				counter: AtomicUsize::new(0),
+				pending: self.pending.clone(),
+			})),
+			// Should not be reachable
+			None => (),
+		}
+		Ok(())
+	}
+    fn on_message(&mut self, msg: Message) -> WsResult<()> {
+		match parse_response(&msg.to_string()) {
+			(Some(id), response) => {
+				match self.pending.remove(id) {
+					Some(c) => c.complete(response),
+					None => println!("warning: unexpected id: {}", id),
+				}
+			}
+			(None, response) => println!("warning: error: {:?}, {}", response, msg.to_string()),
+		}
+		Ok(())
+    }
+}
+
+/// Keeping track of issued requests to be matched up with responses
+#[derive(Clone)]
+struct Pending(Arc<Mutex<BTreeMap<usize, Complete<Result<JsonValue, RpcError>>>>>);
+
+impl Pending {
+	fn new() -> Self {
+		Pending(Arc::new(Mutex::new(BTreeMap::new())))
+	}
+	fn insert(&mut self, k: usize, v: Complete<Result<JsonValue, RpcError>>) {
+		self.0.lock().unwrap().insert(k, v);
```

Comment:
Because this is cloned between the rpc handler inside the websocket event loop and the rpc handler that you get back with Rpc::new()